src/aleph/vm/orchestrator/tasks.py

```diff
@@ -127,3 +134,51 @@ async def stop_watch_for_messages_task(app: web.Application):
         await app["messages_listener"]
     except asyncio.CancelledError:
         logger.debug("Task messages_listener is cancelled now")
+
+
+async def monitor_payments(app: web.Application):
+    logger.debug("Monitoring balances")
+    pool: VmPool = app["vm_pool"]
+    while True:
+        await asyncio.sleep(settings.PAYMENT_MONITOR_INTERVAL)
+
+        # Check if the balance held in the wallet is sufficient holder tier resources
+        for sender, chains in pool.get_executions_by_sender(payment_type=PaymentType.hold).items():
+            for chain, executions in chains.items():
+                balance = await get_balance(sender)
+
+                # Stop executions until the required balance is reached
+                required_balance = get_required_balance(executions)
+                while balance < required_balance:
+                    last_execution = executions.pop(-1)
+                    logger.debug(f"Stopping {last_execution} due to insufficient stream")
+                    await last_execution.stop()
+                    required_balance = get_required_balance(executions)
+
+        # Check if the balance held in the wallet is sufficient stream tier resources
+        for sender, chains in pool.get_executions_by_sender(payment_type=PaymentType.superfluid).items():
```

Comment:
why not use `.is_stream()` instead in `.get_executions_by_sender()`?

Reply:
This is the loop to get the executions by sender and by chain, and the `is_stream` is a property of the payment field on the message, so it's not related.

Reply:
Yes, but my point was to use `is_stream` inside of `get_executuons_by_sender`, instead of passing the `PaymentType` as a parameter. Something like:

```py
get_executions_by_sender(is_stream=True)
```

Reply:
Sorry, but I have to insist on that one. If we bring Solana support in the next months, we will have to do this anyways, as we won't use "superfluid" there, but some other protocol, which will then also return `True` on `execution.is_stream()`.

Reply:
Let's refactor that later then ;-)